R/convergence.R

```diff
@@ -316,6 +316,144 @@ quantile2 <- function(x, probs = c(0.05, 0.95), names = TRUE, ...) {
+#' set: giving the ratio of accuracy to predicting a chain uniformly at random.
+#'
+#' @param x a `draws_df` object or one coercible to a `draws_df` object.
+#'
+#' @param split_chains a Boolean indicating whether to split chains into two equal
+#' halves.
+#'
+#' @param uncertainty whether to provide a vector of R* values (if true) or a single
+#' value (if false).
+#'
+#' @param method machine learning classifer available in caret R package.
+#'
+#' @param hyperparameters hyperparameter settings for classifier given as a list.
+#'
+#' @param nsim number of R* values in returned vector if uncertainty is true.
+#'
+#' @param training_proportion proportion of iterations used to train classifier.
+#'
+#' @details `rstar` provides a measure of MCMC convergence based on whether it is
+#' possible to determine the chain that generated a particular draw with a probability
+#' greater than chance. To do so, it fits a machine learning classifier to a subset of
+#' the original MCMC draws (the training set) and evaluates its predictive accuracy on
+#' the remaining draws (the testing set). If predictive accuracy exceeds chance (i.e.
+#' predicting the chain that generated a draw uniformly at random), the diagnostic
+#' measure, R* > 1, indicating that convergence has yet to occur.
+#'
+#' The statistic, R*, is stochastic, meaning that each time the test is run, unless
+#' the random seed is fixed, it will generally produce a different result. To minimise
+#' the implications of this stochasticity, it is recommended to repeatedly run this
+#' function to calculate a distribution of R*; alternatively, an approximation to this
+#' distribution can be obtained by setting uncertainty = T.
+#'
+#' By default, a random forest classifier is used, which tends to perform best for
+#' target distribution dimensionalities of around 4 and above. For lower dimensional
+#' targets, gradient boosted models (called via `method=\"gbm\"`) tend to have a
+#' higher classification accuracy. On a given MCMC sample, it is recommended to
+#' try both of these classifiers.
+#'
+#' @return A single numeric R* value (by default) or a vector of values (if
+#' uncertainty is true).
+#'
+#' @references Ben Lambert, Aki Vehtari (2020) R*: A robust MCMC convergence
+#' diagnostic with uncertainty using gradient-boosted machines
+#' \emph{arXiv preprint} \code{arXiv:TBD}
+#'
+#' @examples
+#' x <- example_draws("eight_schools")
+#' rstar(x)
+#' rstar(x, split_chains = F)
+#' rstar(x, method = "gbm")
+#' # can pass additional arguments to methods
+#' rstar(x, method = "gbm", verbose = F)
+#'
+#' # with uncertainty, returns a vector of R* values
+#' hist(rstar(x, uncertainty = T))
+#' hist(rstar(x, uncertainty = T, nsim = 100))
+#'
+#' # can use other classification methods in caret library
+#' rstar(x, method = "knn")
+#'
+#' @export
+rstar <- function(x, split_chains=T, uncertainty=F, method=NULL, hyperparameters=NULL,
```

Comment:
Shall we make `rstar` generic with corresponding methods? @jgabry what do you think?